File: benchmark/recursive-tree/tree.lua
local Node = {}
Node.__index = Node

function Node:new(value)
    local node = {
        value = value,
        left = nil,
        right = nil
    }
    setmetatable(node, Node)
    return node
end

function Node:insert(val)
    if val < self.value then
        if self.left == nil then
            self.left = Node:new(val)
        else
            self.left:insert(val)
        end
    else
        if self.right == nil then
            self.right = Node:new(val)
        else
            self.right:insert(val)
        end
    end
end

function Node:sum()
    local total = self.value
    if self.left ~= nil then
        total = total + self.left:sum()
    end
    if self.right ~= nil then
        total = total + self.right:sum()
    end
    return total
end

function Node:count()
    local c = 1
    if self.left ~= nil then
        c = c + self.left:count()
    end
    if self.right ~= nil then
        c = c + self.right:count()
    end
    return c
end

function Node:height()
    local lh = 0
    local rh = 0
    if self.left ~= nil then
        lh = self.left:height()
    end
    if self.right ~= nil then
        rh = self.right:height()
    end
    if lh > rh then
        return lh + 1
    else
        return rh + 1
    end
end

local function makeTree(values)
    if #values == 0 then
        return nil
    end
    local root = Node:new(values[1])
    for i = 2, #values do
        root:insert(values[i])
    end
    return root
end

local function fibonacci(n)
    if n <= 1 then
        return n
    end
    return fibonacci(n - 1) + fibonacci(n - 2)
end

local function processWithClosure(arr, multiplier)
    local result = {}
    local fib10 = fibonacci(10)
    
    local function transform(x)
        return x * multiplier + fib10
    end
    
    for i = 1, #arr do
        result[i] = transform(arr[i])
    end
    
    return result
end

local function matrixMultiply(a, b)
    local n = #a
    local result = {}
    for i = 1, n do
        result[i] = {}
        for j = 1, n do
            local sum = 0
            for k = 1, n do
                sum = sum + a[i][k] * b[k][j]
            end
            result[i][j] = sum
        end
    end
    return result
end

local function main()
    local iterations = 500
    local treeSize = 100
    local matrixSize = 20
    
    local totalSum = 0
    local totalCount = 0
    local totalHeight = 0
    
    for iter = 0, iterations - 1 do
        local values = {}
        for i = 0, treeSize - 1 do
            values[i + 1] = (i * 7 + iter * 13) % 1000
        end
        
        local tree = makeTree(values)
        totalSum = totalSum + tree:sum()
        totalCount = totalCount + tree:count()
        totalHeight = totalHeight + tree:height()
        
        local processed = processWithClosure(values, iter % 10 + 1)
        for i = 1, #processed do
            totalSum = totalSum + (processed[i] % 100)
        end
        
        if iter % 50 == 0 then
            local matA = {}
            local matB = {}
            for i = 1, matrixSize do
                matA[i] = {}
                matB[i] = {}
                for j = 1, matrixSize do
                    matA[i][j] = (i - 1 + j - 1 + iter) % 10
                    matB[i][j] = ((i - 1) * (j - 1) + iter) % 10
                end
            end
            
            local matC = matrixMultiply(matA, matB)
            for i = 1, matrixSize do
                for j = 1, matrixSize do
                    totalSum = totalSum + matC[i][j]
                end
            end
        end
    end
    
    print(totalSum, totalCount, totalHeight)
end

main()

File: benchmark/recursive-tree/tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
    
    def insert(self, val):
        if val < self.value:
            if self.left is None:
                self.left = Node(val)
            else:
                self.left.insert(val)
        else:
            if self.right is None:
                self.right = Node(val)
            else:
                self.right.insert(val)
    
    def sum(self):
        total = self.value
        if self.left is not None:
            total += self.left.sum()
        if self.right is not None:
            total += self.right.sum()
        return total
    
    def count(self):
        c = 1
        if self.left is not None:
            c += self.left.count()
        if self.right is not None:
            c += self.right.count()
        return c
    
    def height(self):
        lh = 0
        rh = 0
        if self.left is not None:
            lh = self.left.height()
        if self.right is not None:
            rh = self.right.height()
        return max(lh, rh) + 1

def make_tree(values):
    if len(values) == 0:
        return None
    root = Node(values[0])
    for i in range(1, len(values)):
        root.insert(values[i])
    return root

def fibonacci(n):
    if n <= 1:
        return n
    return fibonacci(n - 1) + fibonacci(n - 2)

def process_with_closure(arr, multiplier):
    fib10 = fibonacci(10)
    
    def transform(x):
        return x * multiplier + fib10
    
    return [transform(x) for x in arr]

def matrix_multiply(a, b):
    n = len(a)
    result = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            sum_val = 0
            for k in range(n):
                sum_val += a[i][k] * b[k][j]
            result[i][j] = sum_val
    return result

def main():
    iterations = 500
    tree_size = 100
    matrix_size = 20
    
    total_sum = 0
    total_count = 0
    total_height = 0
    
    for iter in range(iterations):
        values = [(i * 7 + iter * 13) % 1000 for i in range(tree_size)]
        
        tree = make_tree(values)
        total_sum += tree.sum()
        total_count += tree.count()
        total_height += tree.height()
        
        processed = process_with_closure(values, iter % 10 + 1)
        for val in processed:
            total_sum += val % 100
        
        if iter % 50 == 0:
            mat_a = [[(i + j + iter) % 10 for j in range(matrix_size)] for i in range(matrix_size)]
            mat_b = [[(i * j + iter) % 10 for j in range(matrix_size)] for i in range(matrix_size)]
            
            mat_c = matrix_multiply(mat_a, mat_b)
            for i in range(matrix_size):
                for j in range(matrix_size):
                    total_sum += mat_c[i][j]
    
    print(total_sum, total_count, total_height)

if __name__ == '__main__':
    main()
